src/toil/lib/plugins.py

```diff
@@ -0,0 +1,66 @@
+import importlib
+from typing import Any, Literal
+import pkgutil
+from toil.lib.memoize import memoize
+
+
+PluginType = Literal["batch_system"] | Literal["url_access"]
+plugin_types: list[PluginType] = ["batch_system", "url_access"]
+
+_registry: dict[str, dict[str, Any]] = {k: {} for k in plugin_types}
+
+def register_plugin(
+    plugin_type: PluginType, plugin_name: str, plugin_being_registered: Any
+) -> None:
+    """
+    Adds a plugin to the registry for the given type of plugin.
```

Comment:
Since the type annotations don't define it, it might be good to note here what the expected plugin values are meant to be for a given plugin type (i.e. functions that return a class implementing blah interface).

It also might be good to explain what `plugin_name` is used for: for patch systems it is the value the user passes to `--batchSystem` to actually use it, and for URLs it is the URL scheme value that the plugin gets called to handle. It's not *really* the "name of the plugin" like the variable name would suggest; you shouldn't pass `"My Cool Plugin"` or something.